packages/data/src/namespace-store/index.js

```diff
@@ -27,29 +36,46 @@ export default function createNamespace( key, options, registry ) {
 	const reducer = options.reducer;
 	const store = createReduxStore( key, options, registry );
 
-	let selectors, actions, resolvers;
-	if ( options.actions ) {
-		actions = mapActions( options.actions, store );
-	}
-	if ( options.selectors ) {
-		selectors = mapSelectors( options.selectors, store, registry );
-	}
+	let resolvers;
+	const actions = mapActions( {
+		...metadataActions,
+		...options.actions,
+	}, store );
+	let selectors = mapSelectors( {
+		...mapValues( metadataSelectors, ( selector ) => ( state, ...args ) => selector( state.metadata, ...args ) ),
+		...mapValues( options.selectors, ( selector ) => {
+			if ( selector.isRegistrySelector ) {
```

Comment:
This logic occurs for every selector call? In addition to the overhead of [`mapSelectors`](https://github.com/WordPress/gutenberg/blob/8243a1c9f6561c2dcbd0874349e044a5467c9321/packages/data/src/namespace-store.js#L111)? Should expect a negative performance impact?

Reply:
I wonder if it's not unreasonable to consider/implement what we have currently as `metadata` and `root` as in-fact two separate stores, so that we don't have to override the selector call to reach into state to provide the corrected "root" state.

Reply:
to your second point, in theory, a store could be just a scalar value which would break this proposal.

to your first point, this only happens when registering the store which I guess is fine. (only run once)

Reply:
> to your first point, this only happens when registering the store which I guess is fine. (only run once)

Subsequent calls would still be subject to going through the `( state, ...args ) => selector( state.root, ...args )` proxying? May be small overhead; curious though given the reality of many thousands of selector calls during typical use.